# IMDB Data

In [1]:
import pandas as pd
import numpy as np

In [2]:
basics = pd.read_csv("https://datasets.imdbws.com/title.basics.tsv.gz", sep='\t',
                    low_memory=False)

In [3]:
ratings = pd.read_csv("https://datasets.imdbws.com/title.ratings.tsv.gz",
                      sep='\t', low_memory=False)

In [4]:
akas = pd.read_csv("https://datasets.imdbws.com/title.akas.tsv.gz", sep='\t',
                    low_memory=False)

Some columns need to be converted to numerical

## Cleaning Data

### akas

In [5]:
akas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34526615 entries, 0 to 34526614
Data columns (total 8 columns):
 #   Column           Dtype 
---  ------           ----- 
 0   titleId          object
 1   ordering         int64 
 2   title            object
 3   region           object
 4   language         object
 5   types            object
 6   attributes       object
 7   isOriginalTitle  object
dtypes: int64(1), object(7)
memory usage: 2.1+ GB


Changing \N to NaNs

In [6]:
akas.replace({'\\N': np.nan}, inplace=True)

Keep only US movies

In [7]:
akas['region'].value_counts()

DE    4121666
JP    4119608
FR    4119330
IN    4057408
ES    4041959
       ...   
CC          1
TV          1
NU          1
PW          1
NR          1
Name: region, Length: 247, dtype: int64

In [8]:
akas = akas[akas['region'] == 'US']

In [9]:
akas['region'].value_counts()

US    1405206
Name: region, dtype: int64

In [10]:
akas.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1405206 entries, 5 to 34526359
Data columns (total 8 columns):
 #   Column           Non-Null Count    Dtype 
---  ------           --------------    ----- 
 0   titleId          1405206 non-null  object
 1   ordering         1405206 non-null  int64 
 2   title            1405206 non-null  object
 3   region           1405206 non-null  object
 4   language         3768 non-null     object
 5   types            969941 non-null   object
 6   attributes       45688 non-null    object
 7   isOriginalTitle  1403861 non-null  object
dtypes: int64(1), object(7)
memory usage: 96.5+ MB


### basics

In [11]:
basics.info()
basics.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9523140 entries, 0 to 9523139
Data columns (total 9 columns):
 #   Column          Dtype 
---  ------          ----- 
 0   tconst          object
 1   titleType       object
 2   primaryTitle    object
 3   originalTitle   object
 4   isAdult         object
 5   startYear       object
 6   endYear         object
 7   runtimeMinutes  object
 8   genres          object
dtypes: object(9)
memory usage: 653.9+ MB


tconst titleType            primaryTitle           originalTitle  \
0  tt0000001     short              Carmencita              Carmencita   
1  tt0000002     short  Le clown et ses chiens  Le clown et ses chiens   
2  tt0000003     short          Pauvre Pierrot          Pauvre Pierrot   
3  tt0000004     short             Un bon bock             Un bon bock   
4  tt0000005     short        Blacksmith Scene        Blacksmith Scene   

  isAdult startYear endYear runtimeMinutes                    genres  
0       0      1894      \N              1         Documentary,Short  
1       0      1892      \N              5           Animation,Short  
2       0      1892      \N              4  Animation,Comedy,Romance  
3       0      1892      \N             12           Animation,Short  
4       0      1893      \N              1              Comedy,Short

replace \N with np.nan

In [12]:
basics.replace({'\\N': np.nan}, inplace=True)

Remove NaNs from genre and runtimeMinutes columns

In [13]:
basics['genres'].isna().sum()

434435

In [14]:
basics['runtimeMinutes'].isna().sum()

6769506

In [15]:
basics.dropna(subset=['genres', 'runtimeMinutes'], inplace=True)

In [16]:
basics['genres'].isna().sum()

0

In [17]:
basics['runtimeMinutes'].isna().sum()

0

filter for only movies

In [18]:
basics['titleType'].value_counts()

tvEpisode       1314572
short            588054
movie            374953
video            178152
tvMovie           90497
tvSeries          88683
tvSpecial         17471
tvMiniSeries      16709
tvShort            8618
videoGame           314
Name: titleType, dtype: int64

In [19]:
basics = basics[basics['titleType'] == 'movie']
basics['titleType'].value_counts()

movie    374953
Name: titleType, dtype: int64

Filter for only non-fiction movies

In [20]:
is_doc = basics['genres'].str.contains('documentary',case=False)
basics = basics[~is_doc]

In [21]:
for i in basics['genres']:
    if 'Documentary' in i:
        print(i)
    elif 'documentary' in i:
        print(i)

Keep startYear between and including 2000-2021

In [22]:
basics.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 284931 entries, 8 to 9522990
Data columns (total 9 columns):
 #   Column          Non-Null Count   Dtype 
---  ------          --------------   ----- 
 0   tconst          284931 non-null  object
 1   titleType       284931 non-null  object
 2   primaryTitle    284931 non-null  object
 3   originalTitle   284931 non-null  object
 4   isAdult         284931 non-null  object
 5   startYear       280280 non-null  object
 6   endYear         0 non-null       object
 7   runtimeMinutes  284931 non-null  object
 8   genres          284931 non-null  object
dtypes: object(9)
memory usage: 21.7+ MB


In [23]:
# Shrinking the range of values for releases to movies between
# 2000 up to 2021
basics = basics[basics['startYear'] >= '2000']
basics = basics[basics['startYear'] <= '2021']
print(min(basics['startYear']))
max(basics['startYear'])

2000


'2021'

In [24]:
basics.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 137540 entries, 34803 to 9522990
Data columns (total 9 columns):
 #   Column          Non-Null Count   Dtype 
---  ------          --------------   ----- 
 0   tconst          137540 non-null  object
 1   titleType       137540 non-null  object
 2   primaryTitle    137540 non-null  object
 3   originalTitle   137540 non-null  object
 4   isAdult         137540 non-null  object
 5   startYear       137540 non-null  object
 6   endYear         0 non-null       object
 7   runtimeMinutes  137540 non-null  object
 8   genres          137540 non-null  object
dtypes: object(9)
memory usage: 10.5+ MB


In [25]:
# Removing Empty Column
basics.drop(columns='endYear', inplace=True)
basics.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 137540 entries, 34803 to 9522990
Data columns (total 8 columns):
 #   Column          Non-Null Count   Dtype 
---  ------          --------------   ----- 
 0   tconst          137540 non-null  object
 1   titleType       137540 non-null  object
 2   primaryTitle    137540 non-null  object
 3   originalTitle   137540 non-null  object
 4   isAdult         137540 non-null  object
 5   startYear       137540 non-null  object
 6   runtimeMinutes  137540 non-null  object
 7   genres          137540 non-null  object
dtypes: object(8)
memory usage: 9.4+ MB


Keeping only US movies

In [26]:
basics = basics[basics['tconst'].isin(akas['titleId'])]

In [27]:
basics.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 80431 entries, 34803 to 9522906
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   tconst          80431 non-null  object
 1   titleType       80431 non-null  object
 2   primaryTitle    80431 non-null  object
 3   originalTitle   80431 non-null  object
 4   isAdult         80431 non-null  object
 5   startYear       80431 non-null  object
 6   runtimeMinutes  80431 non-null  object
 7   genres          80431 non-null  object
dtypes: object(8)
memory usage: 5.5+ MB


In [28]:
# Converting necessary columns
# into numerical columns
basics['startYear'] = basics['startYear'].astype(int)
basics['runtimeMinutes'] = basics['runtimeMinutes'].astype(float)
basics['isAdult'] = basics['isAdult'].astype(int)

In [29]:
basics.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 80431 entries, 34803 to 9522906
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   tconst          80431 non-null  object 
 1   titleType       80431 non-null  object 
 2   primaryTitle    80431 non-null  object 
 3   originalTitle   80431 non-null  object 
 4   isAdult         80431 non-null  int32  
 5   startYear       80431 non-null  int32  
 6   runtimeMinutes  80431 non-null  float64
 7   genres          80431 non-null  object 
dtypes: float64(1), int32(2), object(5)
memory usage: 4.9+ MB


### ratings

In [30]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1266231 entries, 0 to 1266230
Data columns (total 3 columns):
 #   Column         Non-Null Count    Dtype  
---  ------         --------------    -----  
 0   tconst         1266231 non-null  object 
 1   averageRating  1266231 non-null  float64
 2   numVotes       1266231 non-null  int64  
dtypes: float64(1), int64(1), object(1)
memory usage: 29.0+ MB


Changing \N to NaNs

In [31]:
ratings.replace({'\\N': np.nan}, inplace=True)

Keep only US movies

In [32]:
ratings = ratings[ratings['tconst'].isin(akas['titleId'])]

In [33]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 484938 entries, 0 to 1266217
Data columns (total 3 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   tconst         484938 non-null  object 
 1   averageRating  484938 non-null  float64
 2   numVotes       484938 non-null  int64  
dtypes: float64(1), int64(1), object(1)
memory usage: 14.8+ MB


## Saving Data

In [34]:
import os
os.makedirs('movie-data/',exist_ok=True) 
# Confirm folder created
os.listdir("movie-data/")

['final_tmdb_data_2000.csv.gz',
 'final_tmdb_data_2001.csv.gz',
 'title_akas.csv.gz',
 'title_basics.csv.gz',
 'title_ratings.csv.gz',
 'tmdb_api_results_Drama.json',
 'tmdb_results_combined.csv.gz']

In [35]:
## Save current dataframe to file.
basics.to_csv("movie-data/title_basics.csv.gz",compression='gzip',index=False)

In [36]:
akas.to_csv("movie-data/title_akas.csv.gz",compression='gzip',index=False)

In [37]:
ratings.to_csv("movie-data/title_ratings.csv.gz",compression='gzip',index=False)

# TMDB API data

## Obtaining Data
___
I need to obtain financial data as well as certification/MPAA ratings.

In [38]:
import json, time, os
from tqdm.notebook import tqdm_notebook

with open('/Users/aveld/secret/tmdb_api.json') as f:
    login = json.load(f)

In [39]:
import tmdbsimple as tmdb

tmdb.API_KEY = login['api-key']

In [40]:
# Function to get MPAA
def movie_rating(id):
    """Adapted from source = https://github.com/celiao/tmdbsimple"""
    
    # Get the movie object for the current id
    movie = tmdb.Movies(id)
    # save the .info .releases dictionaries
    info = movie.info()
    releases = movie.releases()
    # Loop through countries in releases
    for c in releases['countries']:
        # if the country abbreviation==US
        if c['iso_3166_1' ] =='US':
            ## save a "certification" key in the info dict with the certification
            info['certification'] = c['certification']
    return info

In [41]:
# Function to add new data to a JSON file
def write_json(new_data, filename): 
    """Appends a list of records (new_data) to a json file (filename). 
    Adapted from: https://www.geeksforgeeks.org/append-to-json-file-using-python/"""  
    
    with open(filename,'r+') as file:
        # First we load existing data into a dict.
        file_data = json.load(file)
        ## Choose extend or append
        if (type(new_data) == list) & (type(file_data) == list):
            file_data.extend(new_data)
        else:
             file_data.append(new_data)
        # Sets file's current position at offset.
        file.seek(0)
        # convert back to json.
        json.dump(file_data, file)

In [42]:
def create_json_file(JSON_FILE,  delete_if_exists=False):
    
    ## Check if JSON_FILE exists
    file_exists = os.path.isfile(JSON_FILE)
    
    ## If it DOES exist:
    if file_exists == True:
        
        ## Check if user wants to delete JSON file
        if delete_if_exists==True:
            
            print(f"[!] {JSON_FILE} already exists. Deleting previous file...")
            
            ## delete file and confirm it no longer exits.
            os.remove(JSON_FILE)
            
            ## Recursive call to function after old file deleted
            create_json_file(JSON_FILE,delete_if_exists=False)
        else:
            print(f"[i] {JSON_FILE} already exists.")            
            
            
    ## If it does NOT exist:
    else:
        
        ## INFORM USER AND SAVE EMPTY LIST
        print(f"[i] {JSON_FILE} not found. Saving empty list to new file.")
        
        ## CREATE ANY NEEDED FOLDERS
        # Get the Folder Name only
        folder = os.path.dirname(JSON_FILE)
        
        ## If JSON_FILE included a folder:
        if len(folder)>0:
            # create the folder
            os.makedirs(folder,exist_ok=True)
        ## Save empty list to start the json file
        with open(JSON_FILE,'w') as f:
            json.dump([{'imdb_id':0}],f)  

In [43]:
basics = pd.read_csv('movie-data/title_basics.csv.gz')

In [44]:
basics.head()

tconst titleType                                       primaryTitle  \
0  tt0035423     movie                                     Kate & Leopold   
1  tt0062336     movie  The Tango of the Widower and Its Distorting Mi...   
2  tt0069049     movie                         The Other Side of the Wind   
3  tt0088751     movie                                  The Naked Monster   
4  tt0096056     movie                               Crime and Punishment   

                               originalTitle  isAdult  startYear  \
0                             Kate & Leopold        0       2001   
1  El Tango del Viudo y Su Espejo Deformante        0       2020   
2                 The Other Side of the Wind        0       2018   
3                          The Naked Monster        0       2005   
4                       Crime and Punishment        0       2002   

   runtimeMinutes                  genres  
0           118.0  Comedy,Fantasy,Romance  
1            70.0                   Drama  
2           122.0                   Drama  
3           100.0    Comedy,Horror,Sci-Fi  
4           126.0                   Drama

In [45]:
# Lists
YEARS = [2000, 2001]
errors = []

In [46]:
JSON_FILE = f"movie-data/tmdb_api_results_{i}.json"

In [47]:
# Testing functions 1
movie_rating('tt0848228')

{'adult': False,
 'backdrop_path': '/9BBTo63ANSmhC4e6r62OJFuK2GL.jpg',
 'belongs_to_collection': {'id': 86311,
  'name': 'The Avengers Collection',
  'poster_path': '/yFSIUVTCvgYrpalUktulvk3Gi5Y.jpg',
  'backdrop_path': '/zuW6fOiusv4X9nnW3paHGfXcSll.jpg'},
 'budget': 220000000,
 'genres': [{'id': 878, 'name': 'Science Fiction'},
  {'id': 28, 'name': 'Action'},
  {'id': 12, 'name': 'Adventure'}],
 'homepage': 'https://www.marvel.com/movies/the-avengers',
 'id': 24428,
 'imdb_id': 'tt0848228',
 'original_language': 'en',
 'original_title': 'The Avengers',
 'overview': 'When an unexpected enemy emerges and threatens global safety and security, Nick Fury, director of the international peacekeeping agency known as S.H.I.E.L.D., finds himself in need of a team to pull the world back from the brink of disaster. Spanning the globe, a daring recruitment effort begins!',
 'popularity': 126.272,
 'poster_path': '/tYqp6vEOo8YlVWrYQvt9nyOhsA2.jpg',
 'production_companies': [{'id': 420,
   'logo_pat

In [48]:
# Testing Function 2
movie_rating('tt0332280')

{'adult': False,
 'backdrop_path': '/qom1SZSENdmHFNZBXbtJAU0WTlC.jpg',
 'belongs_to_collection': None,
 'budget': 29000000,
 'genres': [{'id': 10749, 'name': 'Romance'}, {'id': 18, 'name': 'Drama'}],
 'homepage': 'http://www.newline.com/properties/notebookthe.html',
 'id': 11036,
 'imdb_id': 'tt0332280',
 'original_language': 'en',
 'original_title': 'The Notebook',
 'overview': "An epic love story centered around an older man who reads aloud to a woman with Alzheimer's. From a faded notebook, the old man's words bring to life the story about a couple who is separated by World War II, and is then passionately reunited, seven years later, after they have taken different paths.",
 'popularity': 52.669,
 'poster_path': '/rNzQyW4f8B8cQeg7Dgj3n6eT5k9.jpg',
 'production_companies': [{'id': 12,
   'logo_path': '/iaYpEp3LQmb8AfAtmTvpqd4149c.png',
   'name': 'New Line Cinema',
   'origin_country': 'US'},
  {'id': 1565, 'logo_path': None, 'name': 'Avery Pix', 'origin_country': 'US'},
  {'id': 26

In [49]:
# Nested Loop for movies starting in 2000 and 2001
for i in tqdm_notebook(YEARS, desc='YEARS', position=0):
    create_json_file(JSON_FILE, delete_if_exists=True)
    
    #Saving new year as the current df
    df = basics.loc[basics['startYear']==i].copy()
    # saving movie ids to list
    movie_ids = df['tconst'].copy()
    
    # Keeping track of previous loops
    previous_df = pd.read_json(JSON_FILE)
    
    # Making sure not to repeat movie IDs
    movie_ids_to_get = movie_ids[~movie_ids.isin(previous_df['imdb_id'])]
    
    for movie_id in tqdm_notebook(movie_ids_to_get,
                                  desc=f'Movies from {i}',
                                  position=1,
                                  leave=True):
        try:
            # Retrieve data for the movie id
            temp = movie_rating(movie_id)  
            # Append/extend results to existing file
            write_json(temp,JSON_FILE)
            # Short 20 ms sleep
            time.sleep(0.02)
            
        except Exception as e:
            errors.append([movie_id, e])
    
    final_year_df = pd.read_json(JSON_FILE)
    final_year_df.to_csv(f"movie-data/final_tmdb_data_{i}.csv.gz",
                         compression="gzip", index=False)


print(f"- Total errors: {len(errors)}")

YEARS:   0%|          | 0/2 [00:00<?, ?it/s]

[!] movie-data/tmdb_api_results_Drama.json already exists. Deleting previous file...
[i] movie-data/tmdb_api_results_Drama.json not found. Saving empty list to new file.


Movies from 2000:   0%|          | 0/1427 [00:00<?, ?it/s]

[!] movie-data/tmdb_api_results_Drama.json already exists. Deleting previous file...
[i] movie-data/tmdb_api_results_Drama.json not found. Saving empty list to new file.


Movies from 2001:   0%|          | 0/1539 [00:00<?, ?it/s]

- Total errors: 454


## Exploratory Data Analysis

In [50]:
df_2000 = pd.read_csv('movie-data/final_tmdb_data_2000.csv.gz')
df_2001 = pd.read_csv('movie-data/final_tmdb_data_2001.csv.gz')

In [51]:
df_years = pd.concat([df_2000, df_2001])
df_years.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2514 entries, 0 to 1295
Data columns (total 26 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   imdb_id                2514 non-null   object 
 1   adult                  2512 non-null   float64
 2   backdrop_path          1331 non-null   object 
 3   belongs_to_collection  201 non-null    object 
 4   budget                 2512 non-null   float64
 5   genres                 2512 non-null   object 
 6   homepage               173 non-null    object 
 7   id                     2512 non-null   float64
 8   original_language      2512 non-null   object 
 9   original_title         2512 non-null   object 
 10  overview               2461 non-null   object 
 11  popularity             2512 non-null   float64
 12  poster_path            2242 non-null   object 
 13  production_companies   2512 non-null   object 
 14  production_countries   2512 non-null   object 
 15  rele

Starting with 2508 total entries

In [52]:
# Dropping rows with 0 in budget and revenue
# f for filtered
df_years_f = df_years[(df_years['budget'] > 0) | (df_years['revenue'] > 0)]

In [53]:
df_years_f.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 626 entries, 1 to 1275
Data columns (total 26 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   imdb_id                626 non-null    object 
 1   adult                  626 non-null    float64
 2   backdrop_path          531 non-null    object 
 3   belongs_to_collection  105 non-null    object 
 4   budget                 626 non-null    float64
 5   genres                 626 non-null    object 
 6   homepage               77 non-null     object 
 7   id                     626 non-null    float64
 8   original_language      626 non-null    object 
 9   original_title         626 non-null    object 
 10  overview               624 non-null    object 
 11  popularity             626 non-null    float64
 12  poster_path            604 non-null    object 
 13  production_companies   626 non-null    object 
 14  production_countries   626 non-null    object 
 15  relea

Q1. How many movies had at least some valid financial information (values > 0 for budget OR revenue)?

> Only 627 entries left after filtering movies with revenue OR budget data.

In [54]:
df_years_f['certification'].value_counts()

R        229
PG-13    131
PG        35
NR        17
G         14
Name: certification, dtype: int64

Q2. How many movies are there in each of the certification categories (G/PG/PG-13/R)?

> - R = 229
> - PG-13 = 131
> - PG = 35
> - NR = 16
> - G = 14

lambda function modified from:
https://re-thought.com/how-to-suppress-scientific-notation-in-pandas/

In [55]:
df_years_f.groupby('certification')['revenue'].mean().apply(lambda x: '%.2f' % x)

certification
G        123746274.93
NR         8887611.47
PG       112649587.00
PG-13     99262990.70
R         33211194.39
Name: revenue, dtype: object

Q3. What is the average revenue per certification category?
>- G = $ 123,746,274.93 

>- NR = $ 9,443,087.19

>- PG = $ 112,357,320.40

>- PG-13 = $ 99,262,990.70

>- R = $ 33,211,194.39


In [56]:
df_years_f.groupby('certification')['budget'].mean().apply(lambda x: '%.2f' % x)

certification
G        40857142.86
NR        6302358.47
PG       44964285.97
PG-13    43153877.31
R        19746744.32
Name: budget, dtype: object

Q4. What is the average budget per certification category?
>- G = $ 40,857,142.86

>- NR = $ 6,693,722.88

>- PG = $ 44,978,571.69

>- PG-13 = $ 43,153,877.31

>- R = $ 19,746,744.32

In [139]:
# Saving Entire Combined Dataframe
df_years_f.to_csv('movie-data/tmdb_results_combined.csv.gz', compression='gzip', index=False)

# Creating SQL Database

## Creating title_genres & genres table

In [141]:
# Load combined dataframe
title_basics_df = pd.read_csv('movie-data/title_basics.csv.gz', compression='gzip')
title_basics_df.head()

tconst titleType                                       primaryTitle  \
0  tt0035423     movie                                     Kate & Leopold   
1  tt0062336     movie  The Tango of the Widower and Its Distorting Mi...   
2  tt0069049     movie                         The Other Side of the Wind   
3  tt0088751     movie                                  The Naked Monster   
4  tt0096056     movie                               Crime and Punishment   

                               originalTitle  isAdult  startYear  \
0                             Kate & Leopold        0       2001   
1  El Tango del Viudo y Su Espejo Deformante        0       2020   
2                 The Other Side of the Wind        0       2018   
3                          The Naked Monster        0       2005   
4                       Crime and Punishment        0       2002   

   runtimeMinutes                  genres  
0           118.0  Comedy,Fantasy,Romance  
1            70.0                   Drama  
2           122.0                   Drama  
3           100.0    Comedy,Horror,Sci-Fi  
4           126.0                   Drama

### List of Unique Genres

In [142]:
# Create a column with a list of each genre
title_basics_df['genres_split'] = title_basics_df['genres'].str.split(',')

# Explode the genres to create a unique row for each genre
exploded_genres = title_basics_df.explode('genres_split')
exploded_genres

tconst titleType                                       primaryTitle  \
0      tt0035423     movie                                     Kate & Leopold   
0      tt0035423     movie                                     Kate & Leopold   
0      tt0035423     movie                                     Kate & Leopold   
1      tt0062336     movie  The Tango of the Widower and Its Distorting Mi...   
2      tt0069049     movie                         The Other Side of the Wind   
...          ...       ...                                                ...   
80429  tt9916190     movie                                          Safeguard   
80429  tt9916190     movie                                          Safeguard   
80429  tt9916190     movie                                          Safeguard   
80430  tt9916362     movie                                              Coven   
80430  tt9916362     movie                                              Coven   

                                   originalTitle  isAdult  startYear  \
0                                 Kate & Leopold        0       2001   
0                                 Kate & Leopold        0       2001   
0                                 Kate & Leopold        0       2001   
1      El Tango del Viudo y Su Espejo Deformante        0       2020   
2                     The Other Side of the Wind        0       2018   
...                                          ...      ...        ...   
80429                                  Safeguard        0       2020   
80429                                  Safeguard        0       2020   
80429                                  Safeguard        0       2020   
80430                                   Akelarre        0       2020   
80430                                   Akelarre        0       2020   

       runtimeMinutes                     genres genres_split  
0               118.0     Comedy,Fantasy,Romance       Comedy  
0               118.0     Comedy,Fantasy,Romance      Fantasy  
0               118.0     Comedy,Fantasy,Romance      Romance  
1                70.0                      Drama        Drama  
2               122.0                      Drama        Drama  
...               ...                        ...          ...  
80429            95.0  Action,Adventure,Thriller       Action  
80429            95.0  Action,Adventure,Thriller    Adventure  
80429            95.0  Action,Adventure,Thriller     Thriller  
80430            92.0              Drama,History        Drama  
80430            92.0              Drama,History      History  

[150631 rows x 9 columns]

In [143]:
# Create np array of each unique genre
split_genres = title_basics_df['genres'].str.split(',')

# Array of unique values
unique_genres = sorted(split_genres.explode().unique())
unique_genres

['Action',
 'Adult',
 'Adventure',
 'Animation',
 'Biography',
 'Comedy',
 'Crime',
 'Drama',
 'Family',
 'Fantasy',
 'Game-Show',
 'History',
 'Horror',
 'Music',
 'Musical',
 'Mystery',
 'News',
 'Reality-TV',
 'Romance',
 'Sci-Fi',
 'Short',
 'Sport',
 'Talk-Show',
 'Thriller',
 'War',
 'Western']

### title_genres Table

In [145]:
title_genres = exploded_genres[['tconst', 'genres_split']].copy()
title_genres

tconst genres_split
0      tt0035423       Comedy
0      tt0035423      Fantasy
0      tt0035423      Romance
1      tt0062336        Drama
2      tt0069049        Drama
...          ...          ...
80429  tt9916190       Action
80429  tt9916190    Adventure
80429  tt9916190     Thriller
80430  tt9916362        Drama
80430  tt9916362      History

[150631 rows x 2 columns]

### Genre Mapper Dictionary

In [146]:
## Making the genre mapper dictionary

# Number of unqiue genres
genre_ints = range(len(unique_genres))

# Dictionary mapping genres as the keys and a number as the value
genre_map = dict(zip(unique_genres, genre_ints))
genre_map

{'Action': 0,
 'Adult': 1,
 'Adventure': 2,
 'Animation': 3,
 'Biography': 4,
 'Comedy': 5,
 'Crime': 6,
 'Drama': 7,
 'Family': 8,
 'Fantasy': 9,
 'Game-Show': 10,
 'History': 11,
 'Horror': 12,
 'Music': 13,
 'Musical': 14,
 'Mystery': 15,
 'News': 16,
 'Reality-TV': 17,
 'Romance': 18,
 'Sci-Fi': 19,
 'Short': 20,
 'Sport': 21,
 'Talk-Show': 22,
 'Thriller': 23,
 'War': 24,
 'Western': 25}

### Replacing string with id in title_genre table

In [148]:
title_genres['genres_id'] = title_genres['genres_split'].map(genre_map)
ttle_genres = title_genres.drop(columns='genres_split', inplace=True)

title_genres

tconst  genres_id
0      tt0035423          5
0      tt0035423          9
0      tt0035423         18
1      tt0062336          7
2      tt0069049          7
...          ...        ...
80429  tt9916190          0
80429  tt9916190          2
80429  tt9916190         23
80430  tt9916362          7
80430  tt9916362         11

[150631 rows x 2 columns]

### Genre Map Table

In [149]:
genre_map_table = pd.DataFrame({'Genre': genre_map.keys(),
                               'Id': genre_map.values()})
genre_map_table

Genre  Id
0       Action   0
1        Adult   1
2    Adventure   2
3    Animation   3
4    Biography   4
5       Comedy   5
6        Crime   6
7        Drama   7
8       Family   8
9      Fantasy   9
10   Game-Show  10
11     History  11
12      Horror  12
13       Music  13
14     Musical  14
15     Mystery  15
16        News  16
17  Reality-TV  17
18     Romance  18
19      Sci-Fi  19
20       Short  20
21       Sport  21
22   Talk-Show  22
23    Thriller  23
24         War  24
25     Western  25

## Creating Database

In [102]:
import pymysql
pymysql.install_as_MySQLdb()

In [104]:
from sqlalchemy import create_engine

In [107]:
from sqlalchemy_utils import create_database, database_exists


connection = "mysql+pymysql://root:root@localhost/movies"
engine = create_engine(connection)

In [108]:
# Create database
create_database(connection)

## Create Tables

### Data to be used

In [121]:
# csv
title_ratings_df = pd.read_csv('movie-data/title_ratings.csv.gz', compression='gzip')

In [150]:
title_basics_df.head()

tconst titleType                                       primaryTitle  \
0  tt0035423     movie                                     Kate & Leopold   
1  tt0062336     movie  The Tango of the Widower and Its Distorting Mi...   
2  tt0069049     movie                         The Other Side of the Wind   
3  tt0088751     movie                                  The Naked Monster   
4  tt0096056     movie                               Crime and Punishment   

                               originalTitle  isAdult  startYear  \
0                             Kate & Leopold        0       2001   
1  El Tango del Viudo y Su Espejo Deformante        0       2020   
2                 The Other Side of the Wind        0       2018   
3                          The Naked Monster        0       2005   
4                       Crime and Punishment        0       2002   

   runtimeMinutes                  genres                genres_split  
0           118.0  Comedy,Fantasy,Romance  [Comedy, Fantasy, Romance]  
1            70.0                   Drama                     [Drama]  
2           122.0                   Drama                     [Drama]  
3           100.0    Comedy,Horror,Sci-Fi    [Comedy, Horror, Sci-Fi]  
4           126.0                   Drama                     [Drama]

In [123]:
title_ratings_df.head()

tconst  averageRating  numVotes
0  tt0000001            5.7      1936
1  tt0000002            5.8       262
2  tt0000005            6.2      2570
3  tt0000006            5.1       177
4  tt0000007            5.4       808

In [136]:
df_years_f.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 626 entries, 1 to 1275
Data columns (total 26 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   imdb_id                626 non-null    object 
 1   adult                  626 non-null    float64
 2   backdrop_path          531 non-null    object 
 3   belongs_to_collection  105 non-null    object 
 4   budget                 626 non-null    float64
 5   genres                 626 non-null    object 
 6   homepage               77 non-null     object 
 7   id                     626 non-null    float64
 8   original_language      626 non-null    object 
 9   original_title         626 non-null    object 
 10  overview               624 non-null    object 
 11  popularity             626 non-null    float64
 12  poster_path            604 non-null    object 
 13  production_companies   626 non-null    object 
 14  production_countries   626 non-null    object 
 15  relea

In [125]:
genre_map_table.head()

Genre  Id
0     Action   0
1      Adult   1
2  Adventure   2
3  Animation   3
4  Biography   4

In [126]:
title_genres.head()

tconst  genres_id
0   tt0035423          5
0   tt0035423          9
0   tt0035423         16
9   tt0114447          0
16  tt0118589          7

### Preparing Data

In [152]:
# Isolating Necessary Columns
title_basics = title_basics_df.drop(columns=['genres', 'genres_split', 'originalTitle',
                                             'isAdult', 'titleType'])
title_basics.head()

tconst                                       primaryTitle  startYear  \
0  tt0035423                                     Kate & Leopold       2001   
1  tt0062336  The Tango of the Widower and Its Distorting Mi...       2020   
2  tt0069049                         The Other Side of the Wind       2018   
3  tt0088751                                  The Naked Monster       2005   
4  tt0096056                               Crime and Punishment       2002   

   runtimeMinutes  
0           118.0  
1            70.0  
2           122.0  
3           100.0  
4           126.0

In [129]:
# title_ratings uses all columns

In [159]:
tmdb_data = df_years_f[['imdb_id', 'revenue', 'budget', 'certification']].copy()

tmdb_data.rename(columns={'imdb_id': 'tconst'}, inplace=True)

tmdb_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 626 entries, 1 to 1275
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   tconst         626 non-null    object 
 1   revenue        626 non-null    float64
 2   budget         626 non-null    float64
 3   certification  426 non-null    object 
dtypes: float64(2), object(2)
memory usage: 40.6+ KB


### Creating Tables

In [154]:
# max string length
max_str_len = title_basics['tconst'].fillna('').map(len).max()

In [156]:
from sqlalchemy.types import *

# Max string lengths for object columns
key = title_basics['tconst'].fillna('').map(len).max()
title = title_basics['primaryTitle'].fillna('').map(len).max()

# Schema dictonary
df_schema = {
    "tconst": String(key+1), 
    "primaryTitle": Text(title+1),
    'startYear':Float(),
    'endYear':Float(),
    'runtimeMinutes':Integer()}

In [161]:
# Creating tables in Database

title_basics.to_sql('title_basics',engine,dtype=df_schema,if_exists='replace',index=False)

title_ratings_df.to_sql('title_ratings',engine,dtype=df_schema,if_exists='replace',index=False)

tmdb_data.to_sql('tmdb_data',engine,dtype=df_schema,if_exists='replace',index=False)

title_genres.to_sql('title_genres',engine,if_exists='replace',index=False)

genre_map_table.to_sql('genres',engine,if_exists='replace',index=False)

26

In [164]:
# Setting Primary Key

engine.execute('ALTER TABLE title_basics ADD PRIMARY KEY (`tconst`);')

In [165]:
engine.execute('ALTER TABLE title_ratings ADD PRIMARY KEY (`tconst`);')

In [166]:
engine.execute('ALTER TABLE tmdb_data ADD PRIMARY KEY (`tconst`);')

###  Checking Tables

In [167]:
f = """SHOW TABLES"""
pd.read_sql(f, engine)

Tables_in_movies
0           genres
1     title_basics
2     title_genres
3    title_ratings
4        tmdb_data

In [169]:
f = """SELECT *
FROM genres
LIMIT 5"""
pd.read_sql(f, engine)

Genre  Id
0     Action   0
1      Adult   1
2  Adventure   2
3  Animation   3
4  Biography   4

In [170]:
f = """SELECT *
FROM title_basics
LIMIT 5"""
pd.read_sql(f, engine)

tconst                                       primaryTitle  startYear  \
0  tt0035423                                     Kate & Leopold     2001.0   
1  tt0062336  The Tango of the Widower and Its Distorting Mi...     2020.0   
2  tt0069049                         The Other Side of the Wind     2018.0   
3  tt0088751                                  The Naked Monster     2005.0   
4  tt0096056                               Crime and Punishment     2002.0   

   runtimeMinutes  
0             118  
1              70  
2             122  
3             100  
4             126

In [171]:
f = """SELECT *
FROM title_ratings
LIMIT 5"""
pd.read_sql(f, engine)

tconst  averageRating  numVotes
0  tt0000001            5.7      1936
1  tt0000002            5.8       262
2  tt0000005            6.2      2570
3  tt0000006            5.1       177
4  tt0000007            5.4       808

In [172]:
f = """SELECT *
FROM title_genres
LIMIT 5"""
pd.read_sql(f, engine)

tconst  genres_id
0  tt0035423          5
1  tt0035423          9
2  tt0035423         18
3  tt0062336          7
4  tt0069049          7

In [173]:
f = """SELECT *
FROM tmdb_data
LIMIT 5"""
pd.read_sql(f, engine)

tconst     revenue      budget certification
0  tt0035423  76019048.0  48000000.0         PG-13
1  tt0113026         0.0  10000000.0          None
2  tt0118589   5271666.0  22000000.0         PG-13
3  tt0118652         0.0   1000000.0             R
4  tt0118694  12854953.0    150000.0            PG